ref/go/src/bech32/bech32.go

```diff
@@ -0,0 +1,203 @@
+			return "", fmt.Errorf("invalid character human-readable part : hrp[%d]=%d", p, c)
+		}
+	}
+	if strings.ToUpper(hrp) != hrp && strings.ToLower(hrp) != hrp {
+		return "", fmt.Errorf("mix case : hrp=%v", hrp)
+	}
+	lower := strings.ToLower(hrp) == hrp
+	hrp = strings.ToLower(hrp)
+	combined := append(data, createChecksum(hrp, data)...)
+	var ret bytes.Buffer
+	ret.WriteString(hrp)
+	ret.WriteString("1")
+	for idx, p := range combined {
+		if int(p) >= len(charset) {
+			return "", fmt.Errorf("invalid data : data[%d]=%d", idx, p)
+		}
+		ret.WriteByte(charset[p])
+	}
+	if lower {
+		return ret.String(), nil
+	}
+	return strings.ToUpper(ret.String()), nil
+}
+
+// Decode decodes bechString(Bech32) returns hrp(human-readable part) and data(32bit data array) / or error
+func Decode(bechString string) (string, []byte, error) {
+	if len(bechString) > 90 {
+		return "", nil, fmt.Errorf("too long : len=%d", len(bechString))
+	}
+	if strings.ToLower(bechString) != bechString && strings.ToUpper(bechString) != bechString {
+		return "", nil, fmt.Errorf("mixed case")
+	}
+	bechString = strings.ToLower(bechString)
+	pos := strings.LastIndex(bechString, "1")
+	if pos < 1 || pos+7 > len(bechString) {
+		return "", nil, fmt.Errorf("separator '1' at invalid position : pos=%d , len=%d", pos, len(bechString))
+	}
+	hrp := bechString[0:pos]
+	for p, c := range hrp {
+		if c < 33 || c > 126 {
+			return "", nil, fmt.Errorf("invalid character human-readable part : bechString[%d]=%d", p, c)
+		}
+	}
+	data := []byte{}
+	for p := pos + 1; p < len(bechString); p++ {
+		d := strings.Index(charset, fmt.Sprintf("%c", bechString[p]))
+		if d == -1 {
+			return "", nil, fmt.Errorf("invalid character data part : bechString[%d]=%d", p, bechString[p])
+		}
+		data = append(data, byte(d))
+	}
+	if !verifyChecksum(hrp, data) {
+		return "", nil, fmt.Errorf("invalid checksum")
+	}
+	return hrp, data[:len(data)-6], nil
+}
+
+// SegwitAddrDecode decodes hrp(human-readable part) Segwit Address(string), returns version(int) and data(bytes array) / or error
+func SegwitAddrDecode(hrp, addr string) (byte, []byte, error) {
+	dechrp, data, err := Decode(addr)
+	if err != nil {
+		return 0, nil, err
+	}
+	if dechrp != hrp {
+		return 0, nil, fmt.Errorf("invalid human-readable part : %s != %s", hrp, dechrp)
+	}
+	if len(data) < 1 {
+		return 0, nil, fmt.Errorf("invalid decode data length : %d", len(data))
+	}
+	if data[0] > 16 {
+		return 0, nil, fmt.Errorf("invalid witness version : %d", data[0])
+	}
+	if (len(data)%8 == 2 && data[len(data)-1]&0x02 != 0) ||
+		(len(data)%8 == 4 && data[len(data)-1]&0x0f != 0) ||
+		(len(data)%8 == 5 && data[len(data)-1]&0x01 != 0) ||
+		(len(data)%8 == 7 && data[len(data)-1]&0x07 != 0) {
+		return 0, nil, fmt.Errorf("illegal zero padding")
+	}
+	base32String := strings.ToLower(addr[(len(dechrp) + 2):(len(addr) - 6)])
+	for len(base32String)%8 != 0 {
+		base32String += "="
+	}
+	res, err := base32Encoding.DecodeString(base32String)
```

Comment:
To be honest, I am unsure what is going on here after you removed `convertbits` and introduced `base32Encoding`. For example, it doesn't look right that you manually base32 decode in `Decode()`, called in the first line of the function, and then repeat the process here somehow.

Manually modifying the encoded data (the manual deleting and modifying of `=` on line 167 and line 193) feels also unclean.

I feel like the previous version with convertbits was clean and understandable. We can possibly make use of the base32 package, but then in a more systematic and cleaner way. 

=> I propose to revert to convertbits for this pull request, finalize review and merge it, and then open a new PR to see if we can get rid of convertbits in a nice way.